### PR TITLE
[opencv.js simd] Make the code work without "__wasm_unimplemented_simd128__"

### PR DESCRIPTION
**Pull request by Wenzhao-Xiang**
fix #287 

### This pullrequest changes

1. remove unimplemented wasm intrin features
2. disable CV_SIMD128_64F

### CODE REVIEW

**Comment by Wenzhao-Xiang** [comment]
@huningxin @terfendail PTAL, Thanks!

**Comment by Wenzhao-Xiang** [comment]
I found 3 kernel test failed. I'll working on this. Please don't merge this PR.

File: modules/core/include/opencv2/core/hal/intrin_wasm.hpp [hunk 1]
```diff
@@ -2763,10 +2763,23 @@ OPENCV_HAL_IMPL_WASM_INITVEC(v_uint16x8, ushort, u16, i16x8, short)
...
 OPENCV_HAL_IMPL_WASM_INITVEC(v_int32x4, int, s32, i32x4, int)
+OPENCV_HAL_IMPL_WASM_INITVEC(v_float32x4, float, f32, f32x4, float)
+
+#ifdef __wasm_unimplemented_simd128__
 OPENCV_HAL_IMPL_WASM_INITVEC(v_uint64x2, uint64, u64, i64x2, int64)
 OPENCV_HAL_IMPL_WASM_INITVEC(v_int64x2, int64, s64, i64x2, int64)
-OPENCV_HAL_IMPL_WASM_INITVEC(v_float32x4, float, f32, f32x4, float)
 OPENCV_HAL_IMPL_WASM_INITVEC(v_float64x2, double, f64, f64x2, double)
+#else
+#define OPENCV_HAL_IMPL_WASM_64_FALLBACK_INITVEC(_Tpvec, _Tp, suffix, _Tps) \
```
Comment (huningxin) [inline modules/core/include/opencv2/core/hal/intrin_wasm.hpp]: Although the usage is for 64F, but the fallback will work for other types. How about naming it as `OPENCV_HAL_IMPL_FALLBACK_INITVEC`?

File: modules/core/include/opencv2/core/hal/intrin_wasm.hpp [hunk 2]
```diff
@@ -2885,18 +2898,28 @@ inline v_int16x8 v_rshr_pack(const v_int32x4& a, const v_int32x4& b)
 template<int n>
 inline v_uint32x4 v_rshr_pack(const v_uint64x2& a, const v_uint64x2& b)
 {
+    #ifdef __wasm_unimplemented_simd128__
```
Comment (huningxin) [inline modules/core/include/opencv2/core/hal/intrin_wasm.hpp]: indent: no space at the beginning of the line?

File: modules/core/include/opencv2/core/hal/intrin_wasm.hpp [hunk 3]
```diff
@@ -2885,18 +2898,28 @@ inline v_int16x8 v_rshr_pack(const v_int32x4& a, const v_int32x4& b)
...
     v128_t delta = wasm_i64x2_splat(((int64)1 << (n-1)));
     v128_t a1 = wasm_u64x2_shr(wasm_i64x2_add(a.val, delta), n);
     v128_t b1 = wasm_u64x2_shr(wasm_i64x2_add(b.val, delta), n);
     return v_uint32x4(wasm_v8x16_shuffle(a1, b1, 0,1,2,3,8,9,10,11,16,17,18,19,24,25,26,27));
+    #else
```
Comment (huningxin) [inline modules/core/include/opencv2/core/hal/intrin_wasm.hpp]: ditto.

File: modules/core/include/opencv2/core/hal/intrin_wasm.hpp [hunk 4]
```diff
@@ -2885,18 +2898,28 @@ inline v_int16x8 v_rshr_pack(const v_int32x4& a, const v_int32x4& b)
...
+    fallback::v_uint64x2 a_(a), b_(b);
+    return fallback::v_rshr_pack<n>(a_, b_);
+    #endif
```
Comment (huningxin) [inline modules/core/include/opencv2/core/hal/intrin_wasm.hpp]: ditto. And following occurrences.

File: modules/core/include/opencv2/core/hal/intrin_wasm.hpp [hunk 5]
```diff
@@ -3240,14 +3273,38 @@ OPENCV_HAL_IMPL_WASM_BIN_OP(-, v_float32x4, wasm_f32x4_sub)
...
 OPENCV_HAL_IMPL_WASM_BIN_OP(+, v_uint64x2, wasm_i64x2_add)
 OPENCV_HAL_IMPL_WASM_BIN_OP(-, v_uint64x2, wasm_i64x2_sub)
 OPENCV_HAL_IMPL_WASM_BIN_OP(+, v_int64x2, wasm_i64x2_add)
 OPENCV_HAL_IMPL_WASM_BIN_OP(-, v_int64x2, wasm_i64x2_sub)
+OPENCV_HAL_IMPL_WASM_BIN_OP(+, v_float64x2, wasm_f64x2_add)
+OPENCV_HAL_IMPL_WASM_BIN_OP(-, v_float64x2, wasm_f64x2_sub)
+OPENCV_HAL_IMPL_WASM_BIN_OP(*, v_float64x2, wasm_f64x2_mul)
+OPENCV_HAL_IMPL_WASM_BIN_OP(/, v_float64x2, wasm_f64x2_div)
+#else
+#define OPENCV_HAL_IMPL_WASM_64_FALLBACK_BIN_OP(bin_op, _Tpvec) \
```
Comment (huningxin) [inline modules/core/include/opencv2/core/hal/intrin_wasm.hpp]: naming as above.

File: modules/core/include/opencv2/core/hal/intrin_wasm.hpp [hunk 6]
```diff
@@ -3240,14 +3273,38 @@ OPENCV_HAL_IMPL_WASM_BIN_OP(-, v_float32x4, wasm_f32x4_sub)
...
+    inline _Tpvec operator bin_op (const _Tpvec& a, const _Tpvec& b) \
+    { \
+        fallback::_Tpvec a_(a), b_(b); \
+        return _Tpvec((a_) bin_op (b_)); \
+    } \
+    inline _Tpvec& operator bin_op##= (_Tpvec& a, const _Tpvec& b) \
+    { \
+        fallback::_Tpvec a_(a), b_(b); \
+        a = _Tpvec((a_) bin_op (b_)); \
```
Comment (huningxin) [inline modules/core/include/opencv2/core/hal/intrin_wasm.hpp]: `a_ bin_op##= b_;`?

File: modules/core/include/opencv2/core/hal/intrin_wasm.hpp [hunk 7]
```diff
@@ -3671,7 +3767,37 @@ inline _Tpsvec v_shr(const _Tpsvec& a) \
...
 OPENCV_HAL_IMPL_WASM_SHIFT_OP(v_uint32x4, v_int32x4, i32x4, u32x4)
+
+#ifdef __wasm_unimplemented_simd128__
 OPENCV_HAL_IMPL_WASM_SHIFT_OP(v_uint64x2, v_int64x2, i64x2, u64x2)
+#else
+#define OPENCV_HAL_IMPL_WASM_64_FALLBACK_SHIFT_OP(_Tpvec) \
+inline _Tpvec operator << (const _Tpvec& a, int imm) \
+{ \
+    fallback::_Tpvec a_(a); \
+    return _Tpvec(a_ << imm); \
```
Comment (huningxin) [inline modules/core/include/opencv2/core/hal/intrin_wasm.hpp]: `return a_ << imm; \` is fine.

**Review by huningxin — COMMENTED** [review]
Finish a review and leave some comments. Please take a look.

File: modules/core/include/opencv2/core/hal/intrin_wasm.hpp (hunk 6 above)
Comment (Wenzhao-Xiang) [reply modules/core/include/opencv2/core/hal/intrin_wasm.hpp]: I think it unnecessary to use it. For example, like `+=`, if I use `a_ bin##= b_`, I should write:
```c++
fallback::_Tpvec a_(a), b_(b); \
a_ bin##= b_; \
a = _Tpvec(a_); \
return a; \
```
Seem more redundant, right?

File: modules/core/include/opencv2/core/hal/intrin_wasm.hpp (hunk 1 above)
Comment (Wenzhao-Xiang) [reply modules/core/include/opencv2/core/hal/intrin_wasm.hpp]: Looks good!

File: modules/core/include/opencv2/core/hal/intrin_wasm.hpp (hunk 2 above)
Comment (Wenzhao-Xiang) [reply modules/core/include/opencv2/core/hal/intrin_wasm.hpp]: Fixed

File: modules/core/include/opencv2/core/hal/intrin_wasm.hpp (hunk 7 above)
Comment (Wenzhao-Xiang) [reply modules/core/include/opencv2/core/hal/intrin_wasm.hpp]: Fixed, thanks

**Comment by Wenzhao-Xiang** [comment]
> I found 3 kernel test failed. I'll working on this. Please don't merge this PR.

It's caused by the optimization of `v_dotprod` https://github.com/huningxin/opencv/pull/285, I'll open another issue to track it.

File: modules/core/include/opencv2/core/hal/intrin_wasm.hpp (hunk 6 above)
Comment (huningxin) [reply modules/core/include/opencv2/core/hal/intrin_wasm.hpp]: The idea is to re-use the same fallback op that is `bin_op##=` here.

Comment (Wenzhao-Xiang) [reply modules/core/include/opencv2/core/hal/intrin_wasm.hpp]: OK! Fixed it.

**Comment by Wenzhao-Xiang** [comment]
@huningxin @terfendail OK to merge. PTAL, thanks!

**Comment by huningxin** [comment]
Looks good. Merge to test.

**Pull request merged**